### PR TITLE
makefiles/kconfig.mk: generate config file from RIOT_CONFIG_% environment variables

### PR DESCRIPTION
**Pull request by miri64**
### Contribution description
If there is one thing I will miss when we go full Kconfig, is setting configuration and module selection via the environment variables without having to touch random files. This drafts out how to do that regardless with Kconfig by using the `RIOT_CONFIG_` prefix for environement variables. At least for the generated output config this seems to be working as I was able to include modules into my build with that, but this is still lacking some functionality:

TODOs:
- [x] Figure out how to include the selection from the environment into `make menuconfig` reliably (sometimes it does already)
- [x] User should get an error when trying to set an invalid value
- [x] Add a check (to `dist/tools/buildsystem_sanity_check`?) to make sure the `RIOT_CONFIG_` environment variables are not used in production code. This is after all only be meant to be used as fluff for devs, Kconfig files should still take precedence.
- [x] Find way to auto-clean config when symbol is removed from environment
- [x] Document this somewhere.

### Testing procedure
I used a random app of those that already support building with Kconfig:

```
$ TEST_KCONFIG=1 make -C tests/driver_dht/ info-modules
make: Entering directory '/home/mlenders/Repositories/RIOT-OS/RIOT/tests/driver_dht'
=== [ATTENTION] Testing Kconfig dependency modelling ===
=== [ATTENTION] Testing Kconfig dependency modelling ===
=== [ATTENTION] Testing Kconfig dependency modelling ===
make: Nothing to be done for 'clean'.
auto_init
auto_init_xtimer
board
core
core_init
core_msg
core_panic
cortexm_common
cortexm_common_periph
cpu
dht
div
fmt
malloc_thread_safe
newlib
newlib_nano
newlib_syscalls_default
periph
periph_common
periph_gpio
periph_init
periph_init_gpio
periph_init_pm
periph_init_timer
periph_init_uart
periph_pm
periph_timer
periph_uart
pm_layered
sam0_common_periph
samd21_vectors
stdio_uart
sys
xtimer
make: Leaving directory '/home/mlenders/Repositories/RIOT-OS/RIOT/tests/driver_dht'
$ TEST_KCONFIG=1 RIOT_CONFIG_MODULE_ZTIMER=y make -C tests/driver_dht/ info-modules
make: Entering directory '/home/mlenders/Repositories/RIOT-OS/RIOT/tests/driver_dht'
=== [ATTENTION] Testing Kconfig dependency modelling ===
=== [ATTENTION] Testing Kconfig dependency modelling ===
=== [ATTENTION] Testing Kconfig dependency modelling ===
make: Nothing to be done for 'clean'.
auto_init
auto_init_xtimer
auto_init_ztimer
board
core
core_init
core_msg
core_panic
cortexm_common
cortexm_common_periph
cpu
dht
div
fmt
frac
malloc_thread_safe
newlib
newlib_nano
newlib_syscalls_default
periph
periph_common
periph_gpio
periph_init
periph_init_gpio
periph_init_pm
periph_init_timer
periph_init_uart
periph_pm
periph_timer
periph_uart
pm_layered
sam0_common_periph
samd21_vectors
stdio_uart
sys
xtimer
ztimer
ztimer_auto_init
ztimer_convert
ztimer_convert_frac
ztimer_convert_shift
ztimer_core
ztimer_extend
make: Leaving directory '/home/mlenders/Repositories/RIOT-OS/RIOT/tests/driver_dht'
```

~~`clean` is used to reset the generated output config, I think? Without it the module list sometimes remains unchanged.~~

Invalid values are pointed out as such:

```
$ RIOT_CI_BUILD=1 TEST_KCONFIG=1 RIOT_CONFIG_MODULE_ZTIMER="foobar" make -C tests/driver_dht/ -j
make: Entering directory '/home/mlenders/Repositories/RIOT-OS/RIOT/tests/driver_dht'
=== [ATTENTION] Testing Kconfig dependency modelling ===
[genconfig.py]:ERROR-Treating Kconfig warnings as errors:
[genconfig.py]:ERROR-=> warning: the value 'foobar' is invalid for MODULE_ZTIMER (defined at /home/mlenders/Repositories/RIOT-OS/RIOT/sys/ztimer/Kconfig:8), which has type bool -- assignment ignored
make: *** No rule to make target '/home/mlenders/Repositories/RIOT-OS/RIOT/tests/driver_dht/bin/samr21-xpro/generated/out.config', needed by 'check-kconfig-errors'.  Stop.
make: *** Waiting for unfinished jobs....
make: Leaving directory '/home/mlenders/Repositories/RIOT-OS/RIOT/tests/driver_dht'
$ RIOT_CI_BUILD=1 TEST_KCONFIG=1 RIOT_CONFIG_MODULE_AT=y make -C tests/driver_dht/ -j
make: Entering directory '/home/mlenders/Repositories/RIOT-OS/RIOT/tests/driver_dht'
=== [ATTENTION] Testing Kconfig dependency modelling ===
[genconfig.py]:ERROR-=> The module MODULE_AT could not be set to y.
[genconfig.py]:ERROR-   Check the following unmet dependencies: MODULE_ISRPIPE (=n), MODULE_ISRPIPE_READ_TIMEOUT (=n)

make: *** No rule to make target '/home/mlenders/Repositories/RIOT-OS/RIOT/tests/driver_dht/bin/samr21-xpro/generated/out.config', needed by 'check-kconfig-errors'.  Stop.
make: *** Waiting for unfinished jobs....
make: Leaving directory '/home/mlenders/Repositories/RIOT-OS/RIOT/tests/driver_dht'
```

### Issues/PRs references
None

### CODE REVIEW

**Comment by leandrolanzieri** [comment]
I would also add to the TODOs list that the user should get an error when trying to set an invalid value (e.g. enabling a module that misses dependencies or a configuration out of range). Currently this fails silently.

https://github.com/RIOT-OS/RIOT/blob/6c93f0e742766e2a35f35e0b6a80d2d25d00791e/dist/tools/kconfiglib/genconfig.py#L235-L243

**Comment by miri64** [comment]
> I would also add to the TODOs list that the user should get an error when trying to set an invalid value (e.g. enabling a module that misses dependencies or a configuration out of range). Currently this fails silently.
> 
> https://github.com/RIOT-OS/RIOT/blob/6c93f0e742766e2a35f35e0b6a80d2d25d00791e/dist/tools/kconfiglib/genconfig.py#L235-L243

Done.

**Comment by miri64** [comment]
> I would also add to the TODOs list that the user should get an error when trying to set an invalid value (e.g. enabling a module that misses dependencies or a configuration out of range). Currently this fails silently.
> 
> https://github.com/RIOT-OS/RIOT/blob/6c93f0e742766e2a35f35e0b6a80d2d25d00791e/dist/tools/kconfiglib/genconfig.py#L235-L243

Any idea why it fails silently?

**Comment by miri64** [comment]
Mhh this seems to be more related to this currently not working without an intermediate `make clean`:

```console
$ RIOT_CI_BUILD=1 TEST_KCONFIG=1 make -C tests/driver_dht/ -j clean
make: Entering directory '/home/mlenders/Repositories/RIOT-OS/RIOT/tests/driver_dht'
=== [ATTENTION] Testing Kconfig dependency modelling ===
make: Leaving directory '/home/mlenders/Repositories/RIOT-OS/RIOT/tests/driver_dht'
$ RIOT_CI_BUILD=1 TEST_KCONFIG=1 make -C tests/driver_dht/ -j
make: Entering directory '/home/mlenders/Repositories/RIOT-OS/RIOT/tests/driver_dht'
=== [ATTENTION] Testing Kconfig dependency modelling ===
=== [ATTENTION] Testing Kconfig dependency modelling ===
Building application "tests_driver_dht" for "samr21-xpro" with MCU "samd21".

   text	   data	    bss	    dec	    hex	filename
  11964	    112	   2340	  14416	   3850	/home/mlenders/Repositories/RIOT-OS/RIOT/tests/driver_dht/bin/samr21-xpro/tests_driver_dht.elf
make: Leaving directory '/home/mlenders/Repositories/RIOT-OS/RIOT/tests/driver_dht'
$ RIOT_CI_BUILD=1 TEST_KCONFIG=1 RIOT_CONFIG_MODULE_AT=1 make -C tests/driver_dht/ -j 
make: Entering directory '/home/mlenders/Repositories/RIOT-OS/RIOT/tests/driver_dht'
=== [ATTENTION] Testing Kconfig dependency modelling ===
Building application "tests_driver_dht" for "samr21-xpro" with MCU "samd21".

   text	   data	    bss	    dec	    hex	filename
  11964	    112	   2340	  14416	   3850	/home/mlenders/Repositories/RIOT-OS/RIOT/tests/driver_dht/bin/samr21-xpro/tests_driver_dht.elf
make: Leaving directory '/home/mlenders/Repositories/RIOT-OS/RIOT/tests/driver_dht'
$ RIOT_CI_BUILD=1 TEST_KCONFIG=1 RIOT_CONFIG_MODULE_AT=1 make -C tests/driver_dht/ -j clean all
make: Entering directory '/home/mlenders/Repositories/RIOT-OS/RIOT/tests/driver_dht'
=== [ATTENTION] Testing Kconfig dependency modelling ===
[genconfig.py]:ERROR-=> The module MODULE_AT could not be set to y.
[genconfig.py]:ERROR-   Check the following unmet dependencies: MODULE_ISRPIPE (=n), MODULE_ISRPIPE_READ_TIMEOUT (=n)

make: Nothing to be done for 'clean'.
make: *** No rule to make target '/home/mlenders/Repositories/RIOT-OS/RIOT/tests/driver_dht/bin/samr21-xpro/generated/out.config', needed by 'check-kconfig-errors'.  Stop.
make: *** Waiting for unfinished jobs....
make: Leaving directory '/home/mlenders/Repositories/RIOT-OS/RIOT/tests/driver_dht'
```

**Comment by miri64** [comment]
Generating an intermediate file based on the `RIOT_CONFIG_%` variables seems to do the trick and making it a dependency of `KCONFIG_OUT_CONFIG` for most of the TODOs.

**Comment by miri64** [comment]
```console
$ RIOT_CI_BUILD=1 TEST_KCONFIG=1 make -C tests/driver_dht/ -j
make: Entering directory '/home/mlenders/Repositories/RIOT-OS/RIOT/tests/driver_dht'
=== [ATTENTION] Testing Kconfig dependency modelling ===
=== [ATTENTION] Testing Kconfig dependency modelling ===
Building application "tests_driver_dht" for "samr21-xpro" with MCU "samd21".

   text	   data	    bss	    dec	    hex	filename
  11964	    112	   2340	  14416	   3850	/home/mlenders/Repositories/RIOT-OS/RIOT/tests/driver_dht/bin/samr21-xpro/tests_driver_dht.elf
make: Leaving directory '/home/mlenders/Repositories/RIOT-OS/RIOT/tests/driver_dht'
$ RIOT_CI_BUILD=1 TEST_KCONFIG=1 RIOT_CONFIG_MODULE_ZTIMER="foobar" make -C tests/driver_dht/ -j
make: Entering directory '/home/mlenders/Repositories/RIOT-OS/RIOT/tests/driver_dht'
=== [ATTENTION] Testing Kconfig dependency modelling ===
[genconfig.py]:ERROR-Treating Kconfig warnings as errors:
[genconfig.py]:ERROR-=> warning: the value 'foobar' is invalid for MODULE_ZTIMER (defined at /home/mlenders/Repositories/RIOT-OS/RIOT/sys/ztimer/Kconfig:8), which has type bool -- assignment ignored
make: *** No rule to make target '/home/mlenders/Repositories/RIOT-OS/RIOT/tests/driver_dht/bin/samr21-xpro/generated/out.config', needed by 'check-kconfig-errors'.  Stop.
make: *** Waiting for unfinished jobs....
make: Leaving directory '/home/mlenders/Repositories/RIOT-OS/RIOT/tests/driver_dht'
$ RIOT_CI_BUILD=1 TEST_KCONFIG=1 RIOT_CONFIG_MODULE_AT=1 make -C tests/driver_dht/ -j
make: Entering directory '/home/mlenders/Repositories/RIOT-OS/RIOT/tests/driver_dht'
=== [ATTENTION] Testing Kconfig dependency modelling ===
[genconfig.py]:ERROR-=> The module MODULE_AT could not be set to y.
[genconfig.py]:ERROR-   Check the following unmet dependencies: MODULE_ISRPIPE (=n), MODULE_ISRPIPE_READ_TIMEOUT (=n)

make: *** No rule to make target '/home/mlenders/Repositories/RIOT-OS/RIOT/tests/driver_dht/bin/samr21-xpro/generated/out.config', needed by 'check-kconfig-errors'.  Stop.
make: *** Waiting for unfinished jobs....
make: Leaving directory '/home/mlenders/Repositories/RIOT-OS/RIOT/tests/driver_dht'
```

**Comment by miri64** [comment]
No longer WIP.

**Comment by miri64** [comment]
> No longer WIP.

As such: rebased and squashed.

**Comment by miri64** [comment]
After some offline discussion with @leandrolanzieri I changed it to only generate an intermediate config file and not to touch `genconfig.py` at all.

**Comment by leandrolanzieri** [comment]
@miri64 do you mind squashing and rebasing?

**Comment by miri64** [comment]
> @miri64 do you mind squashing and rebasing?

Not at all!

**Comment by miri64** [comment]
Ping?

**Comment by miri64** [comment]
Any chance this can go into the current release?

**Comment by miri64** [comment]
Rebased to current master to resolve merge conflict.

**Review by MrKevinWeiss — COMMENTED** [review]
Tested with:
```
RIOT_CONFIG_PKG_UMORSE=n TEST_KCONFIG=1 make menuconfig
```
And it reflects deselected package in menuconfig.

```
RIOT_CONFIG_PKG_UMORSE=n TEST_KCONFIG=1 make menuconfig
```
fails as expected.

- overwriting old values work
- It catches incorrect names



What doesn't work is:

```
make clean
TEST_KCONFIG=1 make menuconfig
make clean
make menuconfig
```

on master this sequence works.  It seems something persists that force modules to be selected but they cannot since TEST_KCONFIG is not selected.

for shell I had:
```
[genconfig.py]:ERROR-=> The module MODULE_SHELL could not be set to y.
[genconfig.py]:ERROR-   Check the following unmet dependencies: TEST_KCONFIG (=n), TEST_KCONFIG (=n)
```
but I think this will be the same for everything.

**Comment by miri64** [comment]
Mhhh but shouldn't `make clean` remove `$(GENERATED_DIR)/env.config`? Not sure where else those settings would come from in your test.

File: makefiles/eclipse.inc.mk [hunk 1]
```diff
@@ -4,6 +4,8 @@
 eclipsesym: $(CURDIR)/eclipsesym.xml
 eclipsesym.xml: eclipsesym
 
+RIOT_CONFIG_MODULE_SHELL=y
```
Comment (miri64) [inline makefiles/eclipse.inc.mk:7]: Could it be this line maybe :grin:? This will be removed. It just here to check the `buildsystem` check extension. And of course this does not work without `TEST_KCONFIG!=1` for the reasons you gave ;-)

**Comment by miri64** [comment]
I removed the test commit a861bdc2c7c72ad5505255cb78cb5b663d6c2878. Now it works:

```console
$ make clean
$ TEST_KCONFIG=1 make menuconfig
=== [ATTENTION] Testing Kconfig dependency modelling ===
=== [ATTENTION] Testing Kconfig dependency modelling ===
Loaded configuration '/home/mlenders/Repositories/RIOT-OS/RIOT/tests/shell/bin/native/generated/out.config'
No changes to save (for '/home/mlenders/Repositories/RIOT-OS/RIOT/tests/shell/bin/native/generated/out.config')
make /home/mlenders/Repositories/RIOT-OS/RIOT/tests/shell/bin/native/generated/autoconf.h
=== [ATTENTION] Testing Kconfig dependency modelling ===
$ make clean
$ make menuconfig
Loaded configuration '/home/mlenders/Repositories/RIOT-OS/RIOT/tests/shell/bin/native/generated/out.config'
No changes to save (for '/home/mlenders/Repositories/RIOT-OS/RIOT/tests/shell/bin/native/generated/out.config')
make /home/mlenders/Repositories/RIOT-OS/RIOT/tests/shell/bin/native/generated/autoconf.h

```

**Comment by MrKevinWeiss** [comment]
Retested and it fixes the problem from https://github.com/RIOT-OS/RIOT/pull/16052#pullrequestreview-692776820

**Review by MrKevinWeiss — APPROVED** [review]
Maybe a bit greedy of me since I need this for my PR, but everything seems to be in order.  ACK.  I will wait a day or so for people to speak up before I click the merge.

File: dist/tools/buildsystem_sanity_check/check.sh [hunk 2]
```diff
@@ -340,10 +340,8 @@ check_no_riot_config() {
 
     patterns+=(-e 'RIOT_CONFIG_.*')
 
-    pathspec+=('Makefile')
-    pathspec+=('Makefile\..\+')
-    pathspec+=('**/Makefile')
-    pathspec+=('**/Makefile\..\+')
+    pathspec+=('Makefile*')
+    pathspec+=('**/Makefile*')
```
Comment (leandrolanzieri) [inline dist/tools/buildsystem_sanity_check/check.sh:344]: Why this change here?

Comment (miri64) [reply dist/tools/buildsystem_sanity_check/check.sh:344]: Arghs, lemme fix that!

Comment (miri64) [reply dist/tools/buildsystem_sanity_check/check.sh:344]: Moved to the right commit :-)'

**Comment by miri64** [comment]
:tada:
